src/pvt.c

```diff
@@ -67,33 +67,28 @@ static void compute_dops(const double H[4][4],
                          const double pos_ecef[3],
                          dops_t *dops)
 {
-  double H_pos_diag[3];
-  double H_ned[3];
-
-  dops->gdop = dops->pdop = dops->tdop = dops->hdop = dops->vdop = 0;
-
   /* PDOP is the norm of the position elements of tr(H) */
-  for (u8 i=0; i<3; i++) {
-    dops->pdop += H[i][i];
-    /* Also get the trace of H position states for use in HDOP/VDOP
-     * calculations.
-     */
-    H_pos_diag[i] = H[i][i];
-  }
-  dops->pdop = sqrt(dops->pdop);
+  double pdop_sq = H[0][0] + H[1][1] + H[2][2];
+  dops->pdop = sqrt(pdop_sq);
 
   /* TDOP is like PDOP but for the time state. */
   dops->tdop = sqrt(H[3][3]);
 
   /* Calculate the GDOP -- ||tr(H)|| = sqrt(PDOP^2 + TDOP^2) */
-  dops->gdop = sqrt(dops->tdop*dops->tdop + dops->pdop*dops->pdop);
-
-  /* HDOP and VDOP are Horizontal and Vertical; we need to rotate the
-   * PDOP into NED frame and then take the separate components.
-   */
-  wgsecef2ned(H_pos_diag, pos_ecef, H_ned);
-  dops->vdop = sqrt(H_ned[2]*H_ned[2]);
-  dops->hdop = sqrt(H_ned[0]*H_ned[0] + H_ned[1]*H_ned[1]);
+  dops->gdop = sqrt(pdop_sq + H[3][3]);
+
+  /* HDOP and VDOP are Horizontal and Vertical; we need to rotate H
+   * into NED frame and then take the separate components. */
+  double M[3][3], Mt[3][3];
+  double H_3x3[3][3], HM[3][3], H_ned[3][3];
+  /* TODO: save a little CPU by taking advantage of symmetry */
```

Comment:
you could also save more CPU by calculating HDOP from VDOP and PDOP:
For v = ecef2n_vector, VDOP^2 = v * H * v^T, which is easy to compute cheaply. Then use HDOP = sqrt(PDOP^2 - VDOP^2). It's probably numerically stable enough too for the values we'll see.